rotkehlchen/chain/ethereum/balancer/balancer.py

```diff
@@ -0,0 +1,321 @@
+            '$token_ids': '[ID!]',
+        }
+        param_values = {
+            'token_ids': token_addresses,
+        }
+        result = self.graph.query(
+            querystr=TOKENPRICES_QUERY.format(),  # ! Remove returns in string
+            param_types=param_types,
+            param_values=param_values,
+        )
+        # ! Checksum address
+        unknown_assets_prices = {
+            to_checksum_address(token['id']): FVal(token['price'])
+            for token in result['tokenPrices']
+        }
+        return unknown_assets_prices
+
+    def _update_assets_prices_in_addresses_balancer_pools(
+            self,
+            addresses_balancer_pools: AddressesBalancerPools,
+            known_assets_prices: AssetsPrices,
+            unknown_assets_prices: AssetsPrices,
+    ) -> AddressesBalancerPools:
+        """Update the pools underlying assets prices in USD (prices obtained
+        via Inquirer and the Balancer subgraph)
+
+        Process highlights
+        ------------------
+        Per each <BalancerPoolAsset> in <BalancerPool>.assets:
+            * Replace <BalancerPoolAsset> including the asset USD price if it
+            exists either in `known_assets_prices` or `unknown_assets_prices`.
+            * Replace <BalancerPool> if any underlying asset has been replaced.
+        """
+        updated_addresses_balancer_pools: AddressesBalancerPools = {}
+        for address, balancer_pools in addresses_balancer_pools.items():
+            updated_balancer_pools: BalancerPools = []
+            for balancer_pool in balancer_pools:
+                updated_balancer_pool_assets: BalancerPoolAssets = []
+                is_updated = False
+                # Try to get price from either known or unknown assets prices.
+                # Otherwise keep existing price (zero)
+                # ! Assets prices values are <FVal> (truthy)
+                for asset in balancer_pool.assets:
+                    if asset.address in known_assets_prices:
+                        asset_usd = known_assets_prices[asset.address]
+                    elif asset.address in unknown_assets_prices:
+                        asset_usd = unknown_assets_prices[asset.address]
+                    else:
+                        asset_usd = ZERO
+
+                    # Replace <BalancerPoolAsset> if asset USD price exists
+                    if asset_usd != ZERO:
+                        updated_balancer_pool_asset = asset._replace(
+                            asset_usd=asset_usd,
+                            user_balance_usd=asset.user_balance * asset_usd,
+                        )
+                        is_updated = True
+                    else:
+                        updated_balancer_pool_asset = asset
+
+                    updated_balancer_pool_assets.append(updated_balancer_pool_asset)
+
+                # Replace <BalancerPool> if any <BalancerPoolAsset> was replaced
+                updated_balancer_pool = (
```

Comment:
Same exact thing here. Don't use an immutable data structure like NamedTuple if you plan to mutate it later. Just use dataclasses here.

Reply:
done